src/main/kotlin/com/sourcegraph/cody/edit/FixupService.kt

```diff
@@ -76,20 +76,17 @@ class FixupService(val project: Project) : Disposable {
 
   fun getActiveSession(): FixupSession? = activeSession
 
-  fun setActiveSession(session: FixupSession) {
-    activeSession?.let { if (it.isShowingAcceptLens()) it.accept() else it.cancel() }
-    waitUntilActiveSessionIsFinished()
-    activeSession = session
-  }
-
-  fun waitUntilActiveSessionIsFinished() {
-    while (activeSession != null) {
-      Thread.sleep(100)
+  fun startNewSession(newSession: FixupSession, code: () -> Unit) {
+    activeSession?.let { currentSession ->
+      if (currentSession.isShowingAcceptLens()) currentSession.accept() else currentSession.cancel()
```

Comment:
My $0.02, let's avoid the situation we had with chat repo picking where Kotlin and TypeScript have separate, uncoordinated representations of the same thing. TypeScript has some facility for multiple simultaneous edits, scheduling idle diff computation work to apply edits, autoaccept, etc. Anything it lacks we should add for the benefit of VSCode and JetBrains. If we want to do bespoke stuff for JetBrains (why?) then we should go do the work in TypeScript to refactor policy so it is out of the way first, like we did for the lens displays.

Reply:
@dominiccooney that change is needed specifically **because** we decided to temporarily diverge from VSCode implementation.
VSCode supports parallel edits, but Jetbrains currently does not. @steveyegge recently cleaned up `FixupService` and removed edits queue to reflect that fact. That is supposed to be temporary situation, and we wan to make it parallel after GA. But for now, if we only support one edit at a time, we need to deal with UI/UX consequences of that.

If we agree on partially blocking user actions based on the active task state (per my question and @danielmarquespt response) I can simplify this code because situation of two simultaneous edits won't be possible.

Reply:
> if we only support one edit at a time

That's the issue, basically... the TypeScript extension supports multiple simultaneous edits, so we're accruing some tech debt here: the TypeScript side can't add anything that would cause simultaneous edits for JetBrains. The rest of the team is about to embark on an effort to increase edits/user-day, and we are surreptitiously adding this subtle constraint on them.

The other way to approach this would be to make a one-edit-at-a-time-mode in the TypeScript extension so when they run into the constraint the abstractions push back on them instead of JetBrains breaking.

I defer to your judgement on this one. Getting edits out from behind the flag to get some real world use and feedback is certainly valuable.

Reply:
> The other way to approach this would be to make a one-edit-at-a-time-mode in the TypeScript extension so when they run into the constraint the abstractions push back on them instead of JetBrains breaking.

I'm worrying that it could turn out to be much more complicated and have much more edge cases than we think.
I do not feel it's worth spending that time a week+ before passing GA release to QA.
I will be happy to revisit it afterwards.